src/Avalonia.Controls/Control.cs

```diff
@@ -215,18 +279,89 @@ void IVisualBrushInitialize.EnsureInitialized()
         /// <returns>The control that receives the focus adorner.</returns>
         protected virtual IControl? GetTemplateFocusTarget() => this;
 
+        /// <summary>
+        /// Invoked as the first step of marking the control as loaded and raising the
+        /// <see cref="Loaded"/> event.
+        /// </summary>
+        internal void OnLoadedCore()
+        {
+            if (_isLoaded == false)
+            {
+                _isLoaded = true;
+                OnLoaded();
+            }
+        }
+
+        /// <summary>
+        /// Invoked as the first step of marking the control as unloaded and raising the
+        /// <see cref="Unloaded"/> event.
+        /// </summary>
+        internal void OnUnloadedCore()
+        {
+            if (_isLoaded)
```

Comment:
Control should be removed from the _loadedQueue here. In case if dettached was called before dispatched. 
Though I am not sure if it will be the case in the real app, but maybe it's the reason of failing tests?

Reply:
Yea, it's good for safety.

Reply:
Updated

Reply:
Since there are two queue's now I'm undoing this change. I don't want to remove from the processing queue once it has started. A new state variable will be added for _isAttachedToVisualTree and checked just before calling the event.